nemo/utils/app_state.py

```diff
@@ -0,0 +1,213 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nemo.utils.metaclasses import Singleton
+
+
+class AppState(metaclass=Singleton):
```

Comment:
What is this used for?

Reply:
It's used for Model Parallelism. For example, MegatronBertEncoder is a neural module, but needs world state information not normally available to neural modules.

Reply:
Also, this is @tkornuta-nvidia implementation from old NeMo, I just copy-pasted it and added the properties I needed.

Reply:
dozens of thumbs up for this :)